client.go

```diff
@@ -0,0 +1,354 @@
+func Watch[T proto.Message](c *ADSClient, name string, watcher Watcher[T]) {
+	if getResourceHandler[T](c).AddWatcher(name, watcher) {
+		c.notifyNewSubscription()
+	}
+}
+
+// getResourceHandler gets or initializes the [internal.ResourceHandler] for the specified type in
+// the given client.
+func getResourceHandler[T proto.Message](c *ADSClient) *internal.ResourceHandler[T] {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	typeURL := utils.GetTypeURL[T]()
+	if hAny, ok := c.handlers[typeURL]; !ok {
+		h := internal.NewResourceHandler[T]()
+		c.handlers[typeURL] = h
+		return h
+	} else {
+		return hAny.(*internal.ResourceHandler[T])
+	}
+}
+
+func (c *ADSClient) getResourceHandler(typeURL string) (internal.RawResourceHandler, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	h, ok := c.handlers[typeURL]
+	return h, ok
+}
+
+// notifyNewSubscription signals to the subscription loop that a new subscription was added.
+func (c *ADSClient) notifyNewSubscription() {
+	select {
+	case c.newSubscription <- struct{}{}:
+	default:
+	}
+}
+
+type subscriptionSet map[string]utils.Set[string]
+
+// getPendingSubscriptions iterates over all the subscriptions returned by invoking
+// [internal.ResourceHandler.AllSubscriptions] on all registered resource handlers, and compares it
+// against the given set of already registered subscriptions. If any are missing, they are added to
+// the returned subscription set after being added to the given set. This means that repeated
+// invocations of this method will return an empty set if no new subscriptions are added in between.
+func (c *ADSClient) getPendingSubscriptions(registeredSubscriptions subscriptionSet) subscriptionSet {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	pendingSubscriptions := make(subscriptionSet)
+	add := func(typeURL string, name string) {
+		registered := internal.GetNestedMap(registeredSubscriptions, typeURL)
+		if !registered.Contains(name) {
+			registered.Add(name)
+			internal.GetNestedMap(pendingSubscriptions, typeURL).Add(name)
+		}
+	}
+
+	for t, handler := range c.handlers {
+		for k := range handler.AllSubscriptions() {
+			add(t, k)
+		}
+	}
+
+	return pendingSubscriptions
+}
+
+// loop simply calls newStream and subscriptionLoop forever, until the underlying gRPC connection is
+// closed.
+func (c *ADSClient) loop() {
+	for {
+		stream, responses, err := c.newStream()
+		if err != nil {
+			return
+		}
+
+		err = c.subscriptionLoop(stream, responses)
+		slog.WarnContext(stream.Context(), "Restarting ADS stream", "err", err)
+	}
+}
+
+// subscriptionLoop is the critical logic loop for the client. It polls the given response channel,
+// notifying watchers when new responses come in. It also waits for any new subscriptions to be
+// registered, and sends them to the server. This returns whenever the stream ends.
+func (c *ADSClient) subscriptionLoop(stream deltaClient, responsesCh <-chan []*ads.DeltaDiscoveryResponse) error {
```

Comment:
a little doc about []*ads.DeltaDiscoveryResponse can help understand that the slice will alway hold response for same resource type. Otherwise, line 227 is very confusing of why only getting the resource handler for the first response.

Reply:
Sure, can do